From 3f42e4a810248e18c9feae5dce0c1b5977ce6096 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Thu, 16 Nov 2023 15:40:10 +0100
Subject: [PATCH 866/985] evl/control: host net-related attributes at 'net'
 factory

We are going to have several more net-related attributes, polluting
the 'control' namespace for this would be wrong.

Create a 'net' factory definining a single device, to which the
net-related attributes are associated.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 include/evl/factory.h |  1 +
 kernel/evl/control.c  | 33 ---------------------------------
 kernel/evl/factory.c  |  3 +++
 kernel/evl/net/init.c | 40 ++++++++++++++++++++++++++++++++++++++++
 4 files changed, 44 insertions(+), 33 deletions(-)

diff --git a/include/evl/factory.h b/include/evl/factory.h
index 2b8f696e3..14ace1510 100644
--- a/include/evl/factory.h
+++ b/include/evl/factory.h
@@ -234,5 +234,6 @@ extern struct evl_factory evl_xbuf_factory;
 extern struct evl_factory evl_proxy_factory;
 extern struct evl_factory evl_observable_factory;
 extern struct evl_factory evl_rng_factory;
+extern struct evl_factory evl_net_factory;
 
 #endif /* !_EVL_FACTORY_H */
diff --git a/kernel/evl/control.c b/kernel/evl/control.c
index 6c27dabe1..2d1b44f3b 100644
--- a/kernel/evl/control.c
+++ b/kernel/evl/control.c
@@ -12,8 +12,6 @@
 #include <evl/factory.h>
 #include <evl/tick.h>
 #include <evl/control.h>
-#include <evl/net/input.h>
-#include <evl/net/skb.h>
 #include <evl/uaccess.h>
 #include <asm/evl/fptest.h>
 
@@ -461,33 +459,6 @@ static DEVICE_ATTR_RO(tp);
 
 #endif
 
-#ifdef CONFIG_EVL_NET
-
-static ssize_t net_vlans_show(struct device *dev,
-			struct device_attribute *attr,
-			char *buf)
-{
-	return evl_net_show_vlans(buf, PAGE_SIZE);
-}
-
-static ssize_t net_vlans_store(struct device *dev,
-			struct device_attribute *attr,
-			const char *buf, size_t count)
-{
-	return evl_net_store_vlans(buf, count);
-}
-static DEVICE_ATTR_RW(net_vlans);
-
-static ssize_t net_clones_show(struct device *dev,
-			struct device_attribute *attr,
-			char *buf)
-{
-	return evl_net_show_clones(buf, PAGE_SIZE);
-}
-static DEVICE_ATTR_RO(net_clones);
-
-#endif
-
 static struct attribute *control_attrs[] = {
 	&dev_attr_state.attr,
 	&dev_attr_abi.attr,
@@ -497,10 +468,6 @@ static struct attribute *control_attrs[] = {
 #endif
 #ifdef CONFIG_EVL_SCHED_TP
 	&dev_attr_tp.attr,
-#endif
-#ifdef CONFIG_EVL_NET
-	&dev_attr_net_vlans.attr,
-	&dev_attr_net_clones.attr,
 #endif
 	NULL,
 };
diff --git a/kernel/evl/factory.c b/kernel/evl/factory.c
index ed5f84e44..a2f87b89f 100644
--- a/kernel/evl/factory.c
+++ b/kernel/evl/factory.c
@@ -48,6 +48,9 @@ static struct evl_factory *factories[] = {
 	&evl_proxy_factory,
 	&evl_observable_factory,
 	&evl_rng_factory,
+#ifdef CONFIG_EVL_NET
+	&evl_net_factory,
+#endif
 #ifdef CONFIG_FTRACE
 	&evl_trace_factory,
 #endif
diff --git a/kernel/evl/net/init.c b/kernel/evl/net/init.c
index 81bec5ba5..695743b92 100644
--- a/kernel/evl/net/init.c
+++ b/kernel/evl/net/init.c
@@ -15,6 +15,7 @@
 #include <evl/net/device.h>
 #include <evl/net/input.h>
 #include <evl/net/output.h>
+#include <evl/net/skb.h>
 #include <evl/net.h>
 
 static struct notifier_block netdev_notifier = {
@@ -75,3 +76,42 @@ void __init evl_net_cleanup(void)
 	evl_net_cleanup_neighbour();
 	evl_net_cleanup_qdisc();
 }
+
+static const struct file_operations net_fops;
+
+static ssize_t vlans_show(struct device *dev,
+			struct device_attribute *attr,
+			char *buf)
+{
+	return evl_net_show_vlans(buf, PAGE_SIZE);
+}
+
+static ssize_t vlans_store(struct device *dev,
+			struct device_attribute *attr,
+			const char *buf, size_t count)
+{
+	return evl_net_store_vlans(buf, count);
+}
+static DEVICE_ATTR_RW(vlans);
+
+static ssize_t clones_show(struct device *dev,
+			struct device_attribute *attr,
+			char *buf)
+{
+	return evl_net_show_clones(buf, PAGE_SIZE);
+}
+static DEVICE_ATTR_RO(clones);
+
+static struct attribute *net_attrs[] = {
+	&dev_attr_vlans.attr,
+	&dev_attr_clones.attr,
+	NULL,
+};
+ATTRIBUTE_GROUPS(net);
+
+struct evl_factory evl_net_factory = {
+	.name	=	"net",
+	.fops	=	&net_fops,
+	.attrs	=	net_groups,
+	.flags	=	EVL_FACTORY_SINGLE,
+};
-- 
2.30.2

